### PR TITLE
Prevent new SpotBugs violations from being introduced

### PR DESCRIPTION
**Pull request by basil**
This PR introduces a ratchet into the build to prevent new SpotBugs violations from being introduced. It doesn't fix the existing violations, but it puts in place a system for triaging and fixing them in the future.

First, I went through any `@SuppressFBWarnings` annotations that didn't have a justification and either added one if I could think of it or otherwise added a justification of "TODO needs triage". This clarifies whether the exclusion is intended to be permanent or temporary.

Next, I removed the custom threshold from `core/pom.xml`. This introduced many new violations into the build. To get the build clean again, I suppressed any new warning with a justification of "TODO needs triage".

The latter would have peppered the codebase with tons of new `@SuppressFBWarnings` annotations, which I deemed too messy. Instead, for the majority of warnings I added an exclusion in `src/spotbugs/spotbugs-excludes.xml` by violation type and class name. I organized that file into two sections: permanent exclusions (at the top) and temporary exclusions (at the bottom), the latter of which are effectively the equivalent of "TODO needs triage" annotations. I couldn't always avoid adding a "TODO needs triage" annotation (for example, if I couldn't easily refer to the class name because it was an anonymous inner class), so I did end up peppering the codebase with a few new `@SuppressFBWarnings` annotations, but not nearly as many as I would have needed to if I didn't use the class approach in `src/spotbugs/spotbugs-excludes.xml`.

The end result is a codebase that is classified into two sections: triaged and untriaged, with a clear process for moving things from untriaged to triaged and a ratchet for ensuring that no new violations are introduced. Folks like @StefanSpieker can pick a class and/or violation to work on, then follow one of two paths:

- If it is a false positive, add a `@SuppressFBWarnings(value = "[…]", justification = "[…]")` annotation indicating the reason why it is a false positive, then remove the exclusion from the temporary exclusions section.
- If it is not a false positive, fix the bug, then remove the exclusion from the temporary exclusions section.

Over time, the temporary exclusions section should shrink as temporary exclusions are classified as either permanent exclusions or bugs to be fixed.

### Proposed changelog entries

N/A

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by StefanSpieker — APPROVED** [review]
I really like the idea of a baseline where we will not get new spotbugs issues added. The exclusion file with as specific as possible exclusions is great and I have no better idea how to achieve this. Thanks for the work! Especially getting rid of unneeded suppressions.

**Review by timja — APPROVED** [review]
Nice! I prefer this much more than the warnings ng approach where it only blocks new issues as you can’t reproduce that locally

**Comment by basil** [comment]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Pull request merged**
